In [62]:
# Librerias 
import numpy as np
import pandas as pd
# import pyfpgrowth
# import time
from efficient_apriori import apriori
# from sklearn import preprocessing
# from scipy import stats

## 1. 
Por cada un de las siguientes preguntas, proveer un ejemplo de una regla de asociación del dominio de “market basket” que
satisface las siguientes condiciones. También, describir si las reglas son interesantes (subjetivamente).

(a) Una regla que que tiene alto soporte y alta confianza

(b) Una regla que tiene razonablemente alto soporte pero baja confianza

(c) Una regla que tiene bajo soporte y baja confianza

(d) Una regla que tiene bajo soporte y alta confianza


(a) Ya que el soporte mide el porcentaje de itemsets en el dataset total, y la confianza mide el porcentaje de los itemsets en el subconjunto de los datos totales que contiene al antecedente. Por lo tanto debe ser una itemset que sea frecuente tanto en el conjunto total de datos como en el subconjunto que contiene el antecedente.
| id |             items             |
|:--:|:-----------------------------:|
|  1 | Pan, huevos, cerveza, gaseosa |
|  2 |  Pan, huevos, cerveza, leche  |
|  3 |  Leche, Pan, huevos, cerveza  |
|  4 |          Pan, leche          |
|  5 |      Pan, huevos, cerveza     |
En este ejemplo : 
$$\{Pan, Huevos\} \Rightarrow Cerveza \\
S = \frac{\sigma(P,H,C)}{|T|}=\frac{4}{5}= 0.8 \\
C = \frac{\sigma(P,H,C)}{\sigma(P,H)}= \frac{4}{4} = 1
$$
(b) 
| id |             items             |
|:--:|:-----------------------------:|
|  1 | Pan, huevos, cerveza, gaseosa |
|  2 |  Pan, huevos, gaseosa, leche  |
|  3 |  Leche, Pan, huevos, cerveza  |
|  4 |          Pan, huevos          |
|  5 |      Pan, huevos, cerveza     |
| 6  | Pan, Huevos, leche            |
| 7  | Leche, pan, gaseosa, huevos   |
| 8  | Huevos, cerveza, pan          |
| 9  | Pan, Huevos, gaseosa          |
| 10 | Pan, leche, gaseosa           |
En este ejemplo : 
$$\{Pan, Huevos\} \Rightarrow Cerveza \\
S = \frac{\sigma(P,H,C)}{|T|}=\frac{3}{10}= 0.3 \\
C = \frac{\sigma(P,H,C)}{\sigma(P,H)}= \frac{3}{8} = 0.375$$
Aunque el soporte siempre será menor o igual que la confianza, ya que en el caso de la confianza es el mismo numerador que el soporte, $\sigma(items)$, dividido en  un número que siempre es menor que $|T|\geq  \sigma(items)$.  el soporte en este caso es de 30%, y la confianza de 0.375, lo cual no es muy alto.

## 2. 
¿Por qué el proceso de descubrimiento de reglas de asociación es relativamente simple comparado con la generación de
grandes conjuntos de ítems en bases de datos transaccionales?

## 3. 
Considere el siguiente conjunto de datos:

| TID |             items            |
|:---:|:----------------------------:|
| T01 |      milk, beer, diapers     |
| T02 |      bread, butter, milk     |
| T03 |    milk, diapers, cookies    |
| T04 |    bread, butter, cookies    |
| T05 |    beer, cookies, diapers    |
| T06 | milk, diapers, bread, butter |
| T07 |    bread, butter, diapers    |
| T08 |         beer, diapers        |
| T09 | milk, diapers, bread, butter |
| T10 |         beer, cookies        |

(a) ¿cuál es el número máximo de reglas de asociación que se pueden generar? (incluyendo reglas con soporte 0)
El número total de asociaciones es de 
$$\sum_{k=1}^{r-1}\left[\binom{r}{k} \times \sum_{j=1}^{r-k}\binom{r-k}{j}\right] = 3^r - 2^{r+1}+1$$
Donde k es el número de items en el antecedente, y j es el número de items en el consecuente. r es el numero total de items posibles. Con esta expresion y con $r=6$, se tiene que $3^6-2^7+1 = 602$ reglas posibles.

(b) ¿cuál es el tamaño máximo de los conjuntos de items frecuentes que se pueden extraer (asumir el umbral de minsoporte
\> 0)?

Con un soporte minimo de 50%, Los items que cumplen este soporte minimo, son ['milk','diapers','bread','butter']. El tamaño máximo de los conjuntos con estos items, es un conjunto con todos los items frecuentes encontrados, osea un conjunto de tamaño 4. 

(c) Escribir una regla que contenga 3 ítems que se genere de este conjunto de datos.

Se puede asignar la regla **bread, butter -> milk**
EL soporte sería $s = \frac{\sigma(milk)}{10}=5/10 = 0.5$, mientras que la confianza es de $c=\frac{\sigma(milk)}{\sigma(bread,butter)}=5/5=1$

(d) Encontrat un conjunto de items (de tamaño mayor a 2) con el valor de soporte máximo.
En base a los items más frecuentes encontrados en el punto b, se arman los grupos:
| **grupo** |           **items**          | **Frecuencia** | **Soporte %** |
|:---------:|:----------------------------:|:--------------:|:-------------:|
|     1     | milk, bread, butter, diapers |        2       |       20      |
|     2     |      milk, bread, butter     |        3       |       30      |
|     3     |     milk,diapers, butter     |        2       |       20      |
|     4     |     milk, bread, diapers     |        2       |       20      |
|     5     |    diapers, bread, butter    |        3       |       30      |

El conjunto de items con mayor soporte de tamaño mayor a dos, puede ser [milk, bread, butter] o [diapers, bread, butter], con un suporte de 30%.

(e) Encontrar un par de ítems (a y b) tal que las reglas a → b y b → a tengan la misma confianza.

La unica manera de que $c=\frac{\sigma(consecuente+antecedente)}{\sigma(antecedente)} =\frac{\sigma(antecedente)}{\sigma(antecedente+consecuente)}$ es si $\sigma(consecuente+antecedente) = {\sigma(antecedente)}  $ es decir si $c = 1$ el conjunto debe aparecer siempre consecuente+antecedente para que esto suceda. En este caso se tiene **butter->bread** con confianza $c = 5/5 =  1$ por lo que es igual a **bread->butter**

## 4. 
Dado la siguiente base de datos X: 

**insertar tabla**

Usando valores de umbral de soporte = 25 % y confianza = 60 %, encuentre:

a) Todos los conjuntos de ítems en la base de datos X

b) Reglas de asociación fuertes para la base de datos X

c) Analice las asociaciones engañosas para el conjunto de reglas obtenido en el numeral anterior.


## 5. 
El algoritmo Apriori usa estrategías de generación y conteo para derivar conjuntos de items frecuentes. Conjuntos de
item de tamaño k + 1 son creados de conjuntos de items de tamaño k. Un conjunto candidato es eliminado si uno de
sus subconjuntos no es frecuente en la fase de poda. Supongamos que el algoritmo Apriori es aplicado a los datos de la
siguiente tabla con un soporte mínimo de 30% (ejercicio realizado en clase).

(a) Dibujar el lattice de los conjuntos de items y etiquetar cada nodo con las siguientes letras:

i. N: si el conjunto no es considerado candidato

ii. F: si el conjunto candidato es frecuente

iii. I: si el conjunto candidato no es frecuente
![Alt text](punto5_a.png)

(b) ¿Cuál es el porcentaje de conjuntos de items frecuente?

El porcentaje es , incluyendo conjuntos de 1 sólo item, $\frac{\text{conjuntos de items frecuentes}}{\text{Total}} = \frac{15}{31} = 0,484 = 48,4\%$

(c) ¿Cuál es el radio de poda en este conjunto de datos? (El radio de poda es definido como el porcentaje de conjuntos de
items no considerados a ser candidatos, ya sea por que no son generados durante la etapa de generación de candidatos,
o por que son podados en la etapa de poda).

Por esta definición, sería el restante de conjuntos es decir,  $\frac{\text{conjuntos de items no candidatos}}{\text{Total}} = \frac{16}{31} = 1- 0,484 = 0,516 = 51,6\%$

(d) ¿Cuál es la rata de falsa alarma? (porcentaje de los conjuntos de items candidatos que son encontrados NO frecuentes
después de calcular el soporte).

Serían $\frac{\text{conjuntos de items no frecuentes}}{\text{Total}} = \frac{5}{31} = 0,161 = 16,1\%$


## 6. 
Dada la siguiente base de datos transaccional Y:

**insertar tabla**

Usar Apriori y FP-Growth para encontrar los conjuntos de ítems frecuentes con mínimo soporte de 2. Tratar con varios
valores de confianza. Ordenar los conjuntos resultados y analizar resultados de cada algoritmo. Comparar eficiencia de los
dos procesos de mineria. Repetir con soporte de 3 y comparar los resultados.

## 7.
Usando el conjunto de datos del punto 5,

(a) realizar la tabla de contingencia para las siguientes reglas: b→c , a→d , b→d , e→c , c→a

Tabla b→c

|           |  $c$ | $\bar{c}$ |    |
|:---------:|:----:|:---------:|:--:|
|     $b$   |   3  |     4     |  7 |
| $\bar{b}$ |   2  |     1     |  3 |
|           |   5  |     5     | 10 |

Tabla a→d

|           | $d$ | $\bar{d}$ |    |
|:---------:|:---:|:---------:|:--:|
|    $a$    |  4  |     1     |  5 |
| $\bar{a}$ |  5  |     0     |  5 |
|           |  9  |     1     | 10 |

Tabla b→d

|           | $d$ | $\bar{d}$ |    |
|:---------:|:---:|:---------:|:--:|
|    $b$    |  6  |     1     |  7 |
| $\bar{b}$ |  3  |     0     |  3 |
|           |  9  |     1     | 10 |

Tabla e→c

|           | $c$ | $\bar{c}$ |    |
|:---------:|:---:|:---------:|:--:|
|    $e$    |  2  |     4     |  6 |
| $\bar{e}$ |  3  |     1     |  4 |
|           |  5  |     1     | 10 |

Tabla c→a

|           | $a$ | $\bar{a}$ |    |
|:---------:|:---:|:---------:|:--:|
|    $c$    |  2  |     3     |  5 |
| $\bar{c}$ |  3  |     2     |  5 |
|           |  5  |     5     | 10 |

(b) usar las tablas de contigencia del punto anterior para computar y “realizar un ranking” de las reglas usando:

i. soporte

ii. confianza

iii. lift

|   regla   | soporte | confianza | lift | ranking |
|:---------:|:-------:|:---------:|:----:|:-------:|
|    b→c    |   30%   |   42.8%   | 85.7%|    3    | 
|    a→d    |   40%   |    80%    | 88.9%|    2    | 
|    b→d    |   60%   |   85.7%   | 95.2%|    1    |  
|    e→c    |   20%   |   33.3%   | 66.7%|    5    | 
|    c→a    |   20%   |    40%    |  80% |    4    | 

Para elegir el ranking, primero se miro la regla con mayor soporte, y se miro a demas la confianza. b→d tiene mayor soporte, confianza y lift, en el siguiente en el ranking está a→d que tiene un valor mayor de soporte, confianza y lift comparado con los restantes de la lista no incluidos en el ranking. A continuacion b→c también tiene valores de soporte, confianza y lift mayores, finalmente e→c y c→a tienen un mismo valor de soporte, por lo que se elige c→a como el siguiente en el ranking por tener mayor confianza y lift. 

## 8. 
Importar el conjunto de datos marketBasquet.csv. Extraer reglas de asociación utilizando diferentes valores mínimos de
confíanza. Analice los resultados. Reporte los itemsets frecuentes con sus respectivos valores de soporte. Revise los valores
de soporte reportados por el operador de itemsets frecuentes. Verifíquelos manualmente. ¿Qué problema se presenta?
Extraiga reglas de asociación utilizando diferentes valores mínimos de confíanza. Analice los resultados. Reporte los
itemsets frecuentes con sus respectivos valores de soporte. Analice los resultados y discuta las diferencias con el esquema
que no posee un filtro de atributos.


## 9. 
Importe el conjunto de datos credit-german.csv (repositorio de machinelearning). Discretice los atributos numéricos en
máximo 5 bins de igual tamaño. Aplique el algoritmo de reglas de asociación a este conjunto. Interprete las reglas producidas.
Varíe los valores de soporte y de confianza ¿Qué sucede? Interprete las reglas producidas y escoja las que en su concepto
son las más interesantes para el problema (justifíque).

In [166]:
#importar el dataset
credit = pd.read_csv('./german.data',delimiter=' ',header=None)
#seleccionar solo numericos y copiar dataset para modificarlo
numerical_columns = credit.select_dtypes('int'or'float').columns
credit_discrete = credit.copy()
# crear df para guardar valores de los bins encontrados
bins_values = pd.DataFrame(data=[],columns = ['1','2','3','4','5','6'],index=numerical_columns)
for num_column in numerical_columns:
    # dividir en 5 bins con etiquetas A-numero de columna-bin
    credit_discrete[num_column] , bins = pd.cut(credit[num_column],bins = 5, labels=[f'A{num_column+1}{i+1}' for i in range(5)],retbins=True)
    # guardar extremos de bins en 6 columnas de bin values
    bins_values.loc[num_column] = bins

In [168]:
credit

0   1    2    3     4    5    6   7    8     9   ...    11  12    13  \
0    A11   6  A34  A43  1169  A65  A75   4  A93  A101  ...  A121  67  A143   
1    A12  48  A32  A43  5951  A61  A73   2  A92  A101  ...  A121  22  A143   
2    A14  12  A34  A46  2096  A61  A74   2  A93  A101  ...  A121  49  A143   
3    A11  42  A32  A42  7882  A61  A74   2  A93  A103  ...  A122  45  A143   
4    A11  24  A33  A40  4870  A61  A73   3  A93  A101  ...  A124  53  A143   
..   ...  ..  ...  ...   ...  ...  ...  ..  ...   ...  ...   ...  ..   ...   
995  A14  12  A32  A42  1736  A61  A74   3  A92  A101  ...  A121  31  A143   
996  A11  30  A32  A41  3857  A61  A73   4  A91  A101  ...  A122  40  A143   
997  A14  12  A32  A43   804  A61  A75   4  A93  A101  ...  A123  38  A143   
998  A11  45  A32  A43  1845  A61  A73   4  A93  A101  ...  A124  23  A143   
999  A12  45  A34  A41  4576  A62  A71   3  A93  A101  ...  A123  27  A143   

       14 15    16 17    18    19 20  
0    A152  2  A173  1  A192  A201  1  
1    A152  1  A173  1  A191  A201  2  
2    A152  1  A172  2  A191  A201  1  
3    A153  1  A173  2  A191  A201  1  
4    A153  2  A173  2  A191  A201  2  
..    ... ..   ... ..   ...   ... ..  
995  A152  1  A172  1  A191  A201  1  
996  A152  1  A174  1  A192  A201  1  
997  A152  1  A173  1  A191  A201  1  
998  A153  1  A173  1  A192  A201  2  
999  A152  1  A173  1  A191  A201  1  

[1000 rows x 21 columns]

In [167]:
bins_values

1       2       3        4        5        6
1     3.932    17.6    31.2     44.8     58.4     72.0
4   231.826  3884.8  7519.6  11154.4  14789.2  18424.0
7     0.997     1.6     2.2      2.8      3.4      4.0
10    0.997     1.6     2.2      2.8      3.4      4.0
12   18.944    30.2    41.4     52.6     63.8     75.0
15    0.997     1.6     2.2      2.8      3.4      4.0
17    0.999     1.2     1.4      1.6      1.8      2.0
20    0.999     1.2     1.4      1.6      1.8      2.0

In [61]:
# visualizamos el dataset obtenido
credit_discrete.head()

0    1    2    3    4    5    6    7    8     9   ...    11    12    13  \
0  A11  A21  A34  A43  A51  A65  A75  A85  A93  A101  ...  A121  A135  A143   
1  A12  A24  A32  A43  A52  A61  A73  A82  A92  A101  ...  A121  A131  A143   
2  A14  A21  A34  A46  A51  A61  A74  A82  A93  A101  ...  A121  A133  A143   
3  A11  A23  A32  A42  A53  A61  A74  A82  A93  A103  ...  A122  A133  A143   
4  A11  A22  A33  A40  A52  A61  A73  A84  A93  A101  ...  A124  A134  A143   

     14    15    16    17    18    19    20  
0  A152  A162  A173  A181  A192  A201  A211  
1  A152  A161  A173  A181  A191  A201  A215  
2  A152  A161  A172  A185  A191  A201  A211  
3  A153  A161  A173  A185  A191  A201  A211  
4  A153  A162  A173  A185  A191  A201  A215  

[5 rows x 21 columns]

Se tomaron valores iniciales de soporte de 60% y de confianza minima de 60%

In [178]:
itemsets, rules =  apriori(credit_discrete.to_numpy(),min_support=0.6,min_confidence=0.6)
print('cantidad de reglas obtenidas %2d'%len(rules))

cantidad de reglas obtenidas 96


Podemos ordenar las reglas de mayor a menor soporte/confianza

In [180]:
import re
for rule in sorted(rules, key=lambda rule: rule.confidence,reverse=True):
  if rule.lift>=1:
    # if len(rule.lhs +rule.rhs)>=1:
      if re.search("A21",str(rule.lhs +rule.rhs)):
        print(rule)

{A201, A211} -> {A101} (conf: 0.916, supp: 0.611, lift: 1.010, conv: 1.108)
{A211} -> {A101} (conf: 0.907, supp: 0.635, lift: 1.000, conv: 1.002)
{A101} -> {A211} (conf: 0.700, supp: 0.635, lift: 1.000, conv: 1.000)
{A101} -> {A201, A211} (conf: 0.674, supp: 0.611, lift: 1.010, conv: 1.020)


Se filtran las 44 reglas inicialmente obtenidas a reglas que tengan lift mayor o igual a 1. Luego se toman reglas que contengan a la variable 'A21'  esta es la última columna del dataset, es una variable binaria, en este caso toma valores de 'A211' que equivale a un buen cliente o 'A212' que equivale a un mal cliente. Ya que esta es la variable objetivo, es decir queremos averiguar si dada una informacion financiera, personal y crediticia de un cliente es un buen candidato para un credito. 

La primera regla que se muestra $\text{A201}, \text{A211} \rightarrow \text{A101}$ indica que clientes clasificados como 'buenos' que además son trabajadores extranjeros no tienen co-deudores o garante, la confianza nos indica que 91.6% de los clientes clasificados como 'buenos' que además son trabajadores extranjeros en el dataset no tenian garante, mientras que el soporte nos indica que este patron se presento en un 61.1% de los datos. 

La segunda regla es similar a la primera, por lo que es redundante. indica que clientes clasificados como 'buenos' no tienen co-deudores o garante. Las otras reglas indican que la dirección de las reglas es en ambos sentidos, es decir la regla $\text{A201}, \text{A211} \rightarrow \text{A101}$ también está entre las reglas generadas como, $\text{A201}, \text{A211} \leftarrow \text{A101}$ con una confianza menor.



Para encontrar otras reglas interesantes, se toma un soporte y confianza mayor. Además tomamos reglas que contengan 3 elementos o más. Se toman reglas con una valor de lift mayor o igual a 1.

In [188]:
itemsets, rules =  apriori(credit_discrete.to_numpy(),min_support=0.7,min_confidence=0.8)
print('cantidad de reglas obtenidas %2d'%len(rules))
for rule in sorted(rules, key=lambda rule: rule.confidence,reverse=True):
  if rule.lift>=1:
    if len(rule.lhs +rule.rhs)>=3:
    #   if re.search("A21",str(rule.lhs +rule.rhs)):
        print(rule)

cantidad de reglas obtenidas 20
{A101, A181} -> {A201} (conf: 0.977, supp: 0.749, lift: 1.014, conv: 1.577)
{A101, A143} -> {A201} (conf: 0.968, supp: 0.718, lift: 1.005, conv: 1.144)
{A143, A201} -> {A101} (conf: 0.918, supp: 0.718, lift: 1.012, conv: 1.136)
{A181, A201} -> {A101} (conf: 0.915, supp: 0.749, lift: 1.008, conv: 1.088)
{A181} -> {A101, A201} (conf: 0.886, supp: 0.749, lift: 1.007, conv: 1.056)
{A143} -> {A101, A201} (conf: 0.882, supp: 0.718, lift: 1.002, conv: 1.017)
{A101, A201} -> {A181} (conf: 0.851, supp: 0.749, lift: 1.007, conv: 1.041)
{A101} -> {A181, A201} (conf: 0.826, supp: 0.749, lift: 1.008, conv: 1.039)
{A101, A201} -> {A143} (conf: 0.816, supp: 0.718, lift: 1.002, conv: 1.010)


Las reglas obtenidas están ordenadas en orden de confianza. La primera de las reglas obtenidas nos muestran que clientes que no tienen co-deudores o garantes y que ademas tienen una persona de la que se hacen cargo suelen también trabajar en el exterior. La segunda regla es similar pero se tienen  clientes que no tienen co-deudores o garantes  y no tienen planes de pago a plazos usualmente trabajan en el exterior. La tercera de las reglas nos indica que aquellos que no tienen planes de pago a plazos y trabajan en el exterior usualmente no tienen co-deudores o garantes. EL resto de las reglas involucran permutaciones de estos 3 elementos. 